database/src/pool/postgres.rs

```diff
@@ -1887,6 +1888,82 @@ where
             }
         }
     }
+
+    async fn mark_benchmark_request_as_completed(
+        &self,
+        benchmark_request: &mut BenchmarkRequest,
+    ) -> anyhow::Result<bool> {
+        anyhow::ensure!(
+            benchmark_request.tag().is_some(),
+            "Benchmark request has no tag"
+        );
+        anyhow::ensure!(
+            benchmark_request.status == BenchmarkRequestStatus::InProgress,
+            "Can only mark benchmark request whos status is in_progress as complete"
+        );
+
+        // Find if the benchmark is completed and update it's status to completed
+        // in one SQL block
+        let row = self
+            .conn()
+            .query_opt(
+                "
+                UPDATE
+                    benchmark_request
+                SET
+                    status = $1,
+                    completed_at = NOW()
+                WHERE
+                    banchmark_request.tag = $2
+                    AND benchmark_request.status != $1
+                    AND NOT EXISTS (
+                        SELECT
+                            1
+                        FROM
+                            job_queue
+                        WHERE
+                            request_tag = benchmark_request.tag
+                            AND status NOT IN ($3, $4)
+                    )
+                    AND (
+                        benchmark_request.parent_sha IS NULL
+                        OR NOT EXISTS (
+                            SELECT
+                                1
+                            FROM
+                                job_queue
+                            JOIN
+                                benchmark_request AS parent_request 
+                            ON
+                                parent_request.tag = benchmark_request.parent_tag
```

Comment:
```suggestion

```
Why do we need the `JOIN`? Wouldn't it work the same if we deleted this?

Reply:
We don't need the `JOIN`. I made a bit of a hash of it. We need to say where the tag matches the parent_sha. A bit like in the the where above; 183a7e00d95b373d68ada2f4272605d12080ca08